### PR TITLE
Concept mapping clean

### PR DESCRIPTION
**Pull request by josauder**

### CODE REVIEW

File: mermaidseg/dataset_reconciliation/registry.py [hunk 1]
```diff
@@ -338,12 +338,13 @@ def conceptid2labelid(self) -> dict[int, int] | None:
...
-        if self._concept_matrix is None:
-            return None
-        column_concepts = self._concept_matrix.columns.get_level_values("concept").tolist()
-        result: dict[int, int] = {}
-        for col_ind, concept_name in enumerate(column_concepts):
-            target_id = self.target_label2id.get(concept_name)
-            if target_id is not None:
-                result[col_ind] = target_id
-        return result
+        return None
```
Comment (josauder) [inline mermaidseg/dataset_reconciliation/registry.py:372]: I guess the entire function should be removed?

File: mermaidseg/datasets/coralscapes/README.md [hunk 2]
```diff
@@ -19,4 +19,4 @@ in
 
 The Coralscapes dataset (https://huggingface.co/datasets/EPFL-ECEO/coralscapes) is a general-purpose dense semantic segmentation dataset for coral reefs. Similar in scope and with the same structure as the widely used Cityscapes dataset for urban scene understanding, Coralscapes allows for the benchmarking of semantic segmentation models in a new challenging domain.
 
-The Coralscapes dataset spans 2075 images at 1024×2048px resolution gathered from 35 dive sites in 5 countries in the Red Sea, labeled in a consistent and speculation-free manner containing 174k polygons over 39 benthic classes.
\ No newline at end of file
+The Coralscapes dataset spans 2075 images at 1024×2048px resolution gathered from 35 dive sites in 5 countries in the Red Sea, labeled in a consistent and speculation-free manner containing 174k polygons over 39 benthic classes.
```
Comment (josauder) [inline mermaidseg/datasets/coralscapes/README.md:22]: Not sure what happened here

File: mermaidseg/model/loss.py [hunk 3]
```diff
@@ -103,11 +103,12 @@ def forward(
...
 
-        label_mask = (target_labels > 0).unsqueeze(1)
+        concept_mask = concept_labels==0
+        concept_labels = concept_labels-1
         per_element_loss = self.concept_loss(concept_outputs, concept_labels)
 
-        masked_loss = per_element_loss * label_mask
-        denom = label_mask.sum() * outputs.shape[1] + 1e-8
+        masked_loss = per_element_loss * concept_mask
+        denom = concept_mask.sum() * outputs.shape[1] + 1e-8
```
Comment (josauder) [inline mermaidseg/model/loss.py]: Use concept_loss = (per_element_loss * concept_mask.detach()).mean() instead of explicit .sum() with denom division

File: mermaidseg/dataset_reconciliation/concepts.py [hunk 4]
```diff
@@ -81,38 +111,293 @@ def generate_hierarchy_path(label: str, hierarchy_dict: dict[str, str]) -> list[
...
+        df_id2source = pd.DataFrame(
+            global_id2source.values(), columns=["source_dataset_source", "source_label_class_name"]
+        )
+        df_id2source["global_id"] = global_id2source.keys()
+        df_id2source = df_id2source[
+            ["global_id", "source_dataset_source", "source_label_class_name"]
+        ]
+        
+        ## TODO: Update
+        ## The following are temporary:
```
Comment (josauder) [inline mermaidseg/dataset_reconciliation/concepts.py]: Fix

File: mermaidseg/dataset_reconciliation/concepts.py [hunk 5]
```diff
@@ -81,38 +111,293 @@ def generate_hierarchy_path(label: str, hierarchy_dict: dict[str, str]) -> list[
...
+        if "coralnet" in df_id2source["source_dataset_source"].unique():
+            coralnet_id2label = fetch_coralnet_id2label()
+            df_id2source["source_label_class_name"] = df_id2source["source_label_class_name"].apply(lambda s: coralnet_id2label.get(s, s))
+        df_id2source["source_label_class_name"] = df_id2source["source_label_class_name"].apply(lambda s: s.lower())
+
+        df_mapping = df_mapping.merge(
+            df_id2source, on=["source_label_class_name", "source_dataset_source"], how="right"
+        )
+
+        ### TODO: FIX BUG
```
Comment (josauder) [inline mermaidseg/dataset_reconciliation/concepts.py]: Fix

File: mermaidseg/dataset_reconciliation/concepts.py [hunk 6]
```diff
@@ -8,12 +8,42 @@
...
 import requests
 import torch
 
 
+DEFAULT_CLASS_TO_CONCEPTS_CSV = (
+    Path(__file__).resolve().parents[2] / "configs" / "class_to_concepts.csv"
+)
+
+
+## TODO: Temporary Fix
```
Comment (josauder) [inline mermaidseg/dataset_reconciliation/concepts.py]: fix

File: mermaidseg/dataset_reconciliation/concepts.py [hunk 7]
```diff
@@ -8,12 +8,42 @@
...
 
+from pathlib import Path
+
 import numpy as np
 import pandas as pd
 import requests
 import torch
 
 
+DEFAULT_CLASS_TO_CONCEPTS_CSV = (
```
Comment (josauder) [inline mermaidseg/dataset_reconciliation/concepts.py:18]: should be moved to config and arguments with this as the default, no?

**Comment by ViktorDomazetoski** [comment]
Fixed CoralNet ID related-bugs - https://github.com/data-mermaid/mermaid-segmentation/issues/121

File: mermaidseg/dataset_reconciliation/concepts.py (hunk 5 above)
Comment (ViktorDomazetoski) [reply mermaidseg/dataset_reconciliation/concepts.py]: Need to add missing labels to class_to_concepts.csv

File: mermaidseg/model/loss.py [hunk 8]
```diff
@@ -103,11 +103,12 @@ def forward(
 
         class_loss_value = self.class_loss(outputs, target_labels)
 
-        label_mask = (target_labels > 0).unsqueeze(1)
+        concept_mask = concept_labels==0
```
Comment (josauder) [inline mermaidseg/model/loss.py]: Actually this is a major bug, no? It should be `concept_mask > 0`

File: mermaidseg/dataset_reconciliation/concepts.py [hunk 9]
```diff
@@ -81,38 +99,286 @@ def generate_hierarchy_path(label: str, hierarchy_dict: dict[str, str]) -> list[
...
+        )
+
+    source_to_binary_concepts, binary_mapping_dictionary = initialize_binary_concept_mapping(
+        df_mapping=df_mapping, num_global_source=num_global_source
+    )
+
+    source_to_concepts = torch.cat([source_to_taxonomic_concepts, source_to_binary_concepts], dim=1)
+    concept_value2id = {**taxonomic_mapping_dictionary, **binary_mapping_dictionary}
+
+    if list(concept_value2id.keys())[:7] != TAXONOMIC_CONCEPTS:
```
Comment (ViktorDomazetoski) [inline mermaidseg/dataset_reconciliation/concepts.py]: For the loss/eval calculations I found it simplest if the taxonomic_concepts are at the start (for the offset calculations) so I just check that here. Maybe should be updated to work when we don't have all 7 taxonomic ranks

**Comment by yeelauren** [comment]
## Resolution: 3-PR Chain on Main

The concept-mapping work is complete and merged to `main` via three sequential PRs:

| PR | Scope | Status |
|---|---|---|
| **#134** | reconciliation foundation | ✅ merged |
| **#136** | datasets, model, logger | ✅ merged |
| **#140** | config restructure, multi-dataset training | ✅ merged |

### Highlights
- Fixed all 23 test failures
- Full multi-dataset reconciliation (SourceLabelRegistry, label alignment)
- Split-config design (data, model, training, logger — orthogonal concerns)
- Cleaned up dead code and stale configs

### Follow-Ups (Tracked, Not Blocking)
- **#139** — `log_dataset_statistics` multi-dataset adaptation
- **#135** — latent TypeError in remote label fetchers (landed in #141)
- **#141** — config/registry structural cleanup
- **#142** — update wiki for new split-config layout

The stale branches (`111-config-refactor`, `concept_mapping_clean`, `concept_mapping_clean_bugfixes`) are retired. `concept_mapping_jonathan` (DPT/LoRA fork) remains separate for later integration.
